cpp/inc/bond/ext/grpc/wait_callback.h

```diff
@@ -0,0 +1,180 @@
+                }
+            });
+    }
+
+    /// @brief Implicitly creates a callback.
+    operator CallbackType()
+    {
+        return callback();
+    }
+
+    /// @brief Waits for a callback created by callback() to be
+    /// invoked.
+    void wait() const
+    {
+        _impl->wait();
+    }
+
+    /// @brief Waits at least \p timeout for a callback created by
+    /// callback() to be invoked.
+    ///
+    /// @param timeout the minimum amount of time to wait.
+    ///
+    /// @return \p true if a callback was invoked. \p false if the timeout
+    /// occured.
+    template <typename Rep, typename Period>
+    bool wait(std::chrono::duration<Rep, Period> timeout) const
+    {
+        return _impl->wait(timeout);
+    }
+
+    /// @brief Gets the response.
+    ///
+    /// @warning Blocks until a callback is invoked.
+    const bond::bonded<TResponse>& response() const
+    {
+        return _impl->response();
+    }
+
+    /// @brief Gets the status.
+    ///
+    /// @warning Blocks until a callback is invoked.
+    const grpc::Status& status() const
+    {
+        return _impl->status();
+    }
+
+private:
+    /// The interesting guts of wait_callback. We use an impl class so that
+    /// we can handout weak_ptr instances to be captured by the callbacks
+    /// without forcing clients of wait_callback to deal with creating a
+    /// shared_ptr.
+    class impl
+    {
+    private:
+        /// mutex to lock the shared state
+        mutable std::mutex _m;
+        /// condition variable used to signal anyone waiting
+        mutable std::condition_variable _cv;
+        /// The response and status, but more importantly, doubles as a flag
+        /// indicating whether a callback has been invoked yet. If this is
+        /// empty, no callback has been invoked yet. If non-empty, a
+        /// callback has already been invoked.
+        boost::optional<std::tuple<bond::bonded<TResponse>, grpc::Status>> _results;
+
+    public:
+        impl() = default;
+        impl(const impl&) = delete;
+        impl(impl&&) = delete;
+        impl& operator=(const impl&) = delete;
+        impl& operator=(impl&&) = delete;
+
+        /// Attempts to set _results. Only the first call to set that
+        /// completes will actuall set _results. The rest will do nothing.
+        void set(const bond::bonded<TResponse>& response, const grpc::Status& status)
+        {
+            std::unique_lock<std::mutex> lock(_m);
+            if (!_results)
+            {
+                _results.emplace(std::make_tuple(response, status));
+                lock.unlock();
+                _cv.notify_all();
+            }
+        }
+
+        void wait() const
+        {
+            std::unique_lock<std::mutex> lock(_m);
+            return _cv.wait(lock, [this]() { return static_cast<bool>(_results); });
+        }
+
+        template <typename Rep, typename Period>
+        bool wait(std::chrono::duration<Rep, Period> timeout) const
```

Comment:
I would probably accept a `const duration&` (at least std does like that, must be for a reason).